Mathlib/FieldTheory/SeparableDegree.lean

```diff
@@ -268,31 +249,488 @@ def embProdEmbOfIsAlgebraic [Algebra E K] [IsScalarTower F E K] (halg : Algebra.
+        (minpoly.dvd_map_of_isScalarTower F L x)
+  set M := L⟮x⟯; clear_value M
+  letI : Algebra L M := Subalgebra.algebra M.toSubalgebra
+  letI : Module L M := Algebra.toModule
+  letI : SMul L M := Algebra.toSMul
+  haveI : IsScalarTower F L M := IntermediateField.isScalarTower M
+  rwa [finSepDegree_mul_finSepDegree_of_isAlgebraic F L M (Algebra.IsAlgebraic.of_finite L M),
+    FiniteDimensional.finrank_mul_finrank F L M] at heq
+
+/-- If `E / F` is a finite extension, then its separable degree is equal to its degree if and
+only if it is a separable extension. -/
+theorem finSepDegree_eq_finrank_iff [FiniteDimensional F E] :
+    finSepDegree F E = finrank F E ↔ IsSeparable F E :=
+  ⟨fun heq ↦ ⟨fun x ↦ by
+    have halg := IsAlgebraic.of_finite F x
+    refine (finSepDegree_adjoin_simple_eq_finrank_iff F E x halg).1 <| le_antisymm
+      (finSepDegree_adjoin_simple_le_finrank F E x halg) <| le_of_not_lt fun h ↦ ?_
+    have := Nat.mul_lt_mul h (finSepDegree_le_finrank F⟮x⟯ E) Fin.size_pos'
+    rw [finSepDegree_mul_finSepDegree_of_isAlgebraic F F⟮x⟯ E (Algebra.IsAlgebraic.of_finite _ E),
+      FiniteDimensional.finrank_mul_finrank F F⟮x⟯ E] at this
+    linarith only [heq, this]⟩, fun _ ↦ finSepDegree_eq_finrank_of_isSeparable F E⟩
+
+end Field
+
+namespace IntermediateField
+
+lemma separable_of_mem_isSeparable {L : IntermediateField F E} [IsSeparable F L]
+    {x : E} (h : x ∈ L) : (minpoly F x).Separable := by
+  simpa only [minpoly_eq] using IsSeparable.separable F (K := L) ⟨x, h⟩
+
+/-- `F⟮x⟯ / F` is a separable extension if and only if `x` is a separable element.
+As a consequence, any rational function of `x` is also a separable element. -/
+theorem isSeparable_adjoin_simple_iff_separable {x : E} :
+    IsSeparable F F⟮x⟯ ↔ (minpoly F x).Separable := by
+  refine ⟨fun _ ↦ ?_, fun hsep ↦ ?_⟩
+  · exact separable_of_mem_isSeparable F E <| mem_adjoin_simple_self F x
+  · have h := hsep.isIntegral
+    haveI := adjoin.finiteDimensional h
+    rwa [← finSepDegree_eq_finrank_iff,
+      finSepDegree_adjoin_simple_eq_finrank_iff F E x h.isAlgebraic]
+
+/-- If `x` and `y` are both separable elements, then `F⟮x, y⟯ / F` is a separable extension.
+As a consequence, any rational function of `x` and `y` is also a separable element. -/
+theorem isSeparable_adjoin_two_of_separable {x y : E}
+    (hx : (minpoly F x).Separable) (hy : (minpoly F y).Separable) :
+    IsSeparable F F⟮x, y⟯ := by
+  let L := F⟮x⟯
+  haveI : FiniteDimensional F F⟮x, y⟯ := finiteDimensional_adjoin fun _ ↦ by
+    rintro (rfl | rfl)
+    exacts [hx.isIntegral, hy.isIntegral]
+  rw [← finSepDegree_eq_finrank_iff]
+  have halg' := hy.isIntegral.isAlgebraic.tower_top L
+  have heq : _ * _ = _ * _ := congr_arg₂ (· * ·)
+    (finSepDegree_adjoin_simple_eq_finrank_iff F E x hx.isIntegral.isAlgebraic |>.2 hx)
+    (finSepDegree_adjoin_simple_eq_finrank_iff L E y halg' |>.2
+      (hy.map (f := algebraMap F F⟮x⟯) |>.of_dvd (minpoly.dvd_map_of_isScalarTower F L y)))
```

Comment:
This piece of code appears several times. Is it a good idea to extract it into a new lemma, for example `Polynomial.Separable.map_minpoly`? Namely, `Separable (minpoly F x)` implies `Separable (minpoly E x)`.